### PR TITLE
3 phase high voltage inverter support

### PR DESCRIPTION
**Pull request by rixxxx**
This is tested and working with Deye 20kw 3phase HV inverter(SUN-20K-SG01-HP3-EU-AM2).

### CODE REVIEW

**Comment by rixxxx** [comment]
sorry for the mess I created with the commits while trying to fixup a style issue black didn't like

File: hass-addon-sunsynk-multi/config.yaml [hunk 1]
```diff
@@ -58,7 +58,7 @@ schema:
       MODBUS_ID: int(1,16)
       DONGLE_SERIAL_NUMBER: int?
       PORT: str
-  SENSOR_DEFINITIONS: list(single-phase|three-phase)
+  SENSOR_DEFINITIONS: list(single-phase|three-phase|three-phase-hv)
```
Comment (kellerza) [inline hass-addon-sunsynk-multi/config.yaml:61]: Can you undo this change for now? Then we can release it on edge and test it for a while

(Changing multi's config without updating the version & tagging it causes inconsistencies between the addon's options & the built container)

File: src/sunsynk/definitions3phhv.py [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+PROG4_TIME.max = PROG5_TIME
+PROG5_TIME.max = PROG6_TIME
+PROG6_TIME.max = PROG1_TIME
+
+PROG_CHARGE_OPTIONS = {
+    0: "Off",
+    1: "On",
+}
+
+# PROG_CHARGE_OPTIONS = {
```
Comment (kellerza) [inline src/sunsynk/definitions3phhv.py:291]: Please remove these commented lines

File: src/sunsynk/definitions3phhv.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    NumberRWSensor(169, "Prog4 Capacity", "%", min=BATTERY_LOW_CAP),
+    NumberRWSensor(170, "Prog5 Capacity", "%", min=BATTERY_LOW_CAP),
+    NumberRWSensor(171, "Prog6 Capacity", "%", min=BATTERY_LOW_CAP),
+    SelectRWSensor(172, "Prog1 charge", options=PROG_CHARGE_OPTIONS),
+    SelectRWSensor(173, "Prog2 charge", options=PROG_CHARGE_OPTIONS),
+    SelectRWSensor(174, "Prog3 charge", options=PROG_CHARGE_OPTIONS),
+    SelectRWSensor(175, "Prog4 charge", options=PROG_CHARGE_OPTIONS),
+    SelectRWSensor(176, "Prog5 charge", options=PROG_CHARGE_OPTIONS),
+    SelectRWSensor(177, "Prog6 charge", options=PROG_CHARGE_OPTIONS),
+    # SelectRWSensor(172, "Prog1 charge", options=PROG_CHARGE_OPTIONS, bitmask=0x03),
```
Comment (kellerza) [inline src/sunsynk/definitions3phhv.py:329]: Please remove these comments

File: www/docs/guide/tested-inverters.md [hunk 4]
```diff
@@ -9,6 +9,7 @@ There are several inverters that are rebranded Deye inverters, so you might have
 | Sunsynk 5.5kW  | Hubble AM-2       | beta/all | @kellerza     | BMS485 (top left)        |
 | Sunsynk 8.8kW  | BSL 8.2 kWH       | 0.0.8    | @dirkackerman | RS485 (1 in image below) |
 | Deye  8kW      | Pylontech US3000C | 0.1.3dev | @Kladrie      | RS485 (top left)         |
+| Deye 20kW 3PH HV | Deye BOS-G      | head |   | @rixxxx       | Solarman                 |
```
Comment (kellerza) [inline www/docs/guide/tested-inverters.md:12]: Is this an extra `|` after head?

File: www/docs/reference/definitions.md [hunk 5]
```diff
@@ -51,6 +51,17 @@ These definitions are used when you configure a three-phase inverter in the addo
 ```yaml
 SENSOR_DEFINITIONS: three-phase
 ```
+```
```
Comment (kellerza) [inline www/docs/reference/definitions.md:54]: Extra three backticks?
```suggestion
```

File: src/sunsynk/definitions3phhv.py [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+RATED_POWER = Sensor((20, 21), "Rated power", WATT, 0.1)
+SENSORS += RATED_POWER
+
+SENSORS += (
+    Sensor(
+        0, "Device Type"
+    ),  # {2: "Inverter", 3: "Hybrid Inverter", 4: "Micro Inverter", 5: "3 Phase Hybrid Inverter" }),
+    FaultSensor((555, 556, 557, 558), "Fault"),
+    InverterStateSensor(500, "Overall state"),
+    SDStatusSensor(0, "SD Status", ""),  # type: ignore        # 3 Phase does not have SD Card but crashes when removed
```
Comment (kellerza) [inline src/sunsynk/definitions3phhv.py:166]:
```suggestion
    SDStatusSensor(0, "SD Status", ""),  # type: ignore
```
Not sure why this would crash? Seems like a leftover from the 3-ph

**Review by kellerza — COMMENTED** [review]
Thank you @rixxxx !

This looks really good, some minor comments.

**Comment by rixxxx** [comment]
Thanks for the feedback. I'll incorporate the fixes.

**Comment by rixxxx** [comment]
I'm not 100% sure since I dont have Sofar HYD 10KTL-3PH, but based on the pictures and some register information I found, it seems to be different from Sunsynk/Deye. It doesn't look too difficult to add another variant, but this pull request doesn't address that.

**Comment by RomcaKrestic** [comment]
> I'm not 100% sure since I dont have Sofar HYD 10KTL-3PH, but based on the pictures and some register information I found, it seems to be different from Sunsynk/Deye. It doesn't look too difficult to add another variant, but this pull request doesn't address that.

it is definitely different from Dey, you can see it when sewing the SN inverter. There are some pictures instead of the serial number I entered...
Do you think you would like to deal with it? I would be happy to send you a LOG, or create access to Home Assistant :-)
Thanks for your interest and interesting work for the community

**Comment by rixxxx** [comment]
@kellerza all suggested fixed should be incorporated now. I'm not too familiar how the github ecosystem works and probably created unnecessary churn again, sorry for that.

**Comment by RomcaKrestic** [comment]
I probably won't be able to overcome that, I have no experience with reading data via MPTT and therefore I can't properly edit the definition file in src/sunsync/definitions3phhv.py. My limit is on the LOG file of the MPTT connection to the LSW-3 Logger. In the log, it can be seen that when querying the serial number, some nonsense value is returned and the communication is interrupted.. It looks like the inverter sent non-alphanumeric characters, but some nonsense characters, to the SN query via the Logger.
Would it be possible to somehow display what exactly comes from the inverter during communication, a complete communication LOG?
Thank you, Roman

**Comment by kellerza** [comment]
Thanks @rixxxx - this looks good

I’m not able to test it currently, but will merge - so please give some feedback

**Comment by kellerza** [comment]
@RomcaKrestic you can enable DEBUG to get extra logs

if you can find the registers you can use custome sensors - see mysensors in the docs to create these and test them

**Comment by rixxxx** [comment]
@RomcaKrestic there is nothing common with Deye/Sunsync 3ph hv inverter registers. This probably requires new definition file, doens't make sense to add conditions to each register. I think it would be best if you could file an issue with the problem. I may have a look.

**Comment by rixxxx** [comment]
@kellerza my existing config still works with the merged tree.

**Comment by JanKraslice** [comment]
Hello why even after your update the HV DEYE 30K still shows the battery voltage value for the LV ( low voltage) inverter[

](url)

**Comment by rixxxx** [comment]
@JanKraslice it should be Battery 1 voltage in the latest release code.
image

**Comment by JanKraslice** [comment]
I think I have the latest version, please do I understand correctly that it updates automatically. 

I currently have

Sunsynk/Deye Inverter Add-on (edge/dev)

Current version: c5668d0

**Comment by rixxxx** [comment]
Ok, figured it out. LV definition has "Battery voltage", HV has "Battery 1 voltage. If you use battery_voltage in your configuration it will use the LV variant. I added both battery_1_voltage and battery_voltage and my sensors look like that:
image

I'm not using the HA addon and don't know how to do it but somewhere there should be a list of sensors where you need to change battery_voltage to battery_1_voltage.

**Comment by rixxxx** [comment]
@kellerza do we wan't to break LV backward compatibility and change the sensor name?

**Comment by JanKraslice** [comment]
So in my case I don't know where to change, sorry I guess I need help. thanks (battery_1_voltage.)

**Comment by JanKraslice** [comment]
I've already figured it out and edited it, and you're right. 

battery_1_voltage is o.k.

 

Can you advise how to add a sales disable.

I am talking about the value:

Solar Sell

Max Sell Power

 

In some part of Europe it is not worth selling electricity at certain times according to the exchange prices

Thank you

**Comment by rixxxx** [comment]
Solar Sell should be "solar_export" and Max Sell Power should be "export_limit_power". You need to add them in the same way you did battery_1_voltage

**Comment by JanKraslice** [comment]
Still in the HV version the power sensor is missing the output of both load and backup inverter or one combining both.

The ss_inverter_power is the total power of the inverter both output load to the object and output to the grid which is wrong if sold to the grid.

**Comment by JanKraslice** [comment]
Supplementary essential_power and non_essential_power

15:59:40 INFO Using three-phase HV sensor definitions.

15:59:40 ERROR Unknown sensor entered: aux_power

15:59:40 ERROR Unknown sensor entered: essential_power

15:59:40 ERROR Unknown sensor: grid_voltage

15:59:40 ERROR Neznámné zadané pivot: inverter_current

15:59:40 ERROR Unknown sensor: load_frequency

15:59:40 ERROR non_newsreader: non_essential_power
